components/raftstore/src/store/util.rs

```diff
@@ -1139,16 +1133,24 @@ fn check_remove_or_demote_voter(
     let mut slow_peer_count = 0;
     let mut normal_peer_count = 0;
     // Here we assume if the last beartbeat is within 2 election timeout, the peer
-    // is healthy. This is to be tolerant to some slightly slow peers when
-    // the leader is in hibernate mode.
+    // is healthy. When a region is hibernate, we expect all its peers are *slow*
+    // and it would still allow the operation
     let slow_peer_threshold =
         2 * cfg.raft_base_tick_interval.0 * cfg.raft_max_election_timeout_ticks as u32;
     for (id, last_heartbeat) in peer_heartbeat {
-        // leader itself is not a slow peer
-        if *id == leader_id || last_heartbeat.elapsed() <= slow_peer_threshold {
-            normal_peer_count += 1;
-        } else {
-            slow_peer_count += 1;
+        // for slow and normal peer calculation, we only count voter role
+        if region
+            .get_peers()
+            .iter()
+            .find(|p| p.get_id() == *id)
+            .map_or(false, |p| p.role == PeerRole::Voter)
```

Comment:
What about IncomingVoter and DemotingVoter

Reply:
check_remove_or_demote_voter is called only when it's simple or EnteringJoint, which means the role could never be IncomingVoter or DemotingVoter before the raftcmd is executed. @Connor1996